build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,291 @@
+ * @param currentTypeMap - map containing current type data
+ * @param packageObject - package.json object containing type validation settings
+ * @param testString - array to store generated test strings
+ * @returns - string array representing generated compatibility test cases
+ */
+export function generateCompatibilityTestCases(
+	previousData: TypeData[],
+	currentTypeMap: Map<string, TypeData>,
+	packageObject: PackageJson,
+	testString: string[],
+): string[] {
+	const broken: BrokenCompatTypes = packageObject.typeValidation?.broken ?? {};
+	for (const oldTypeData of previousData) {
+		const oldType: TestCaseTypeData = {
+			prefix: "old",
+			...oldTypeData,
+			removed: false,
+		};
+		const currentTypeData = currentTypeMap.get(getFullTypeName(oldTypeData));
+		// if the current package is missing a type, we will use the old type data.
+		// this can represent a breaking change which can be disable in the package.json.
+		// this can also happen for type changes, like type to interface, which can remain
+		// compatible.
+		const currentType: TestCaseTypeData =
+			currentTypeData === undefined
+				? {
+						prefix: "current",
+						...oldTypeData,
+						kind: `Removed${oldTypeData.kind}`,
+						removed: true,
+					}
+				: {
+						prefix: "current",
+						...currentTypeData,
+						removed: false,
+					};
+
+		// look for settings not under version, then fall back to version for back compat
+		const brokenData = broken?.[getFullTypeName(currentType)];
+
+		testString.push(`/*`);
+		testString.push(`* Validate forward compat by using old type in place of current type`);
+		testString.push(
+			`* If breaking change required, add in package.json under typeValidation.broken:`,
+		);
+		testString.push(`* "${getFullTypeName(currentType)}": {"forwardCompat": false}`);
+		testString.push("*/");
+		testString.push(...buildTestCase(oldType, currentType, brokenData?.forwardCompat ?? true));
+
+		testString.push("");
+
+		testString.push(`/*`);
+		testString.push(`* Validate back compat by using current type in place of old type`);
+		testString.push(
+			`* If breaking change required, add in package.json under typeValidation.broken:`,
+		);
+		testString.push(`* "${getFullTypeName(currentType)}": {"backCompat": false}`);
+		testString.push("*/");
+		testString.push(...buildTestCase(currentType, oldType, brokenData?.backCompat ?? true));
+		testString.push("");
+	}
+	return testString;
+}
+
+/**
+ * Prepares the file path for type validation tests and skips test generation if disabled in package.json.
+ * @param packageObject - the package.json object
+ * @returns type validation file path
+ */
+export function prepareFilepathForTests(packageObject): string {
+	const testPath = typeTestFilePath;
+	// remove scope if it exists
+	const unscopedName = path.basename(packageObject.name);
+
+	const fileBaseName = unscopedName
+		.split("-")
+		.map((p) => p[0].toUpperCase() + p.substring(1))
+		.join("");
```

Comment:
This logic seems worth splitting out into its own function, both because it seems like logic we should reuse, and also because naming the function will make it clearer what this code intends to do.